sphinx/builders/linkcheck.py

```diff
@@ -494,13 +498,23 @@ def check_uri() -> Tuple[str, str, int]:
                 new_url = response.url
                 if anchor:
                     new_url += '#' + anchor
-                # history contains any redirects, get last
-                if response.history:
+
+                if ignored_redirect(req_url, new_url):
+                    return 'working', '', 0
+                elif response.history:
+                    # history contains any redirects, get last
                     code = response.history[-1].status_code
                     return 'redirected', new_url, code
                 else:
                     return 'redirected', new_url, 0
 
+        def ignored_redirect(url: str, new_url: str) -> bool:
+            for from_url, to_url in self.config.linkcheck_ignore_redirects.items():
+                if from_url.match(url) and to_url.match(new_url):
```

Comment:
If `https://www.sphinx-doc.org/en/stable/usage/quickstart.html` redirects to something other than `https://www.sphinx-doc.org/en/master/usage/quickstart.html` then it implies that the page no longer exists in the stable version and it should be possible to consider that an error. The validation should be worth the effort if someone is taking the effort to link to `/stable/` versions (although I'm not linking to readthedocs websites like this myself).

I would be inclined to use `re.sub()` and document it. Alternatively, do the group dict checking but only if the group name exists in both matches?

There may also need to be a regex and non-regex version because the need to escape every `.` for an accurate match may confuse people. Any examples should definitely use `r'...'` strings.

Reply:
> If https://www.sphinx-doc.org/en/stable/usage/quickstart.html redirects to something other than https://www.sphinx-doc.org/en/master/usage/quickstart.html then it implies that the page no longer exists in the stable version and it should be possible to consider that an error.

For a page that does not exist, the server should return a 404 in the first place.

I understand not all web servers behave that way, but let’s keep in mind we’re dealing with poorly configured web servers here. I wouldn’t advise implementing a brittle solution to work around bad behaving servers. I consider the `re.sub()` solution to be brittle:
- disallowing `\` in the `to_url` is sure to trip a good portion of users,
- `re.sub()` was design to perform replacements in a template string, not in a regular expression. There’s no guarantee the regular expression will be valid after the templating. The `\` is a symptom of it, but there may be others.

The comparison of groups seems more robust to me. One can adjust the groups to capture the same portion on both sides. It needs some work to experiment (e.g. try to implement the validation in the #6525 with that system). I’ll try it in the days to come.

> There may also need to be a regex and non-regex version because the need to escape every `.` for an accurate match may confuse people.

I think it would mostly be confusing. There would be two ways of doing the same thing. Besides, there’s a strong probability the unexpected redirect URL will be different enough that the regexp won’t match, even if `r"."` is used instead `r"\."`. Not discussing that `r"\."` is tighter and correct, simply trying to explain why a second setting without regexp seems overkill to me.

Reply:
As @nomis says, the idea here is that it is possible to accept redirect targets matching a certain pattern to be considered "working" and, if they don't match, then they should still show up as "redirected".

I've had a play with the example that @francoisfreitag posted and think that we can make this work by escaping non-backreference escapes:

```python
>>> import re
>>> to_url = r"^https://example.org/\w{2}/\1"
>>> match = re.match(r'https://example\.org/(.*)', 'https://example.org/the-real-page/')
>>> match.expand(to_url)
Traceback (most recent call last):
  File "/usr/lib/python3.9/sre_parse.py", line 1039, in parse_template
    this = chr(ESCAPES[this][1])
KeyError: '\\w'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.9/re.py", line 322, in _expand
    template = sre_parse.parse_template(template, pattern)
  File "/usr/lib/python3.9/sre_parse.py", line 1042, in parse_template
    raise s.error('bad escape %s' % this, len(this))
re.error: bad escape \w at position 21
>>> to_url2 = re.sub(r'(\\[^0-9g])', r'\\\1', to_url)
>>> to_url
'^https://example.org/\\w{2}/\\1'
>>> to_url2
'^https://example.org/\\\\w{2}/\\1'
>>> match.expand(to_url2)
'^https://example.org/\\w{2}/the-real-page/'
```

So `re.sub(r'(\\[^0-9g])', r'\\\1', to_url)` seems to do the trick, escaping non-backreference escapes. It seems that `.expand()` also reverses this extra escaping for us.